skops/io/_general.py

```diff
@@ -242,6 +242,30 @@ def object_get_instance(state, src):
     return instance
 
 
+def method_get_state(obj: Any, save_state: SaveState):
+    # This method is used to persist bound methods, which are
+    # dependent on a specific instance of an object.
+    # It stores the state of the object the method is bound to,
+    # and prepares both to be persisted.
+    res = {
+        "__class__": obj.__class__.__name__,
+        "__module__": get_module(obj),
+        "__loader__": "method_get_instance",
+        "content": {
+            "func": obj.__func__.__name__,
+            "obj": get_state(obj.__self__, save_state),
```

Comment:
I think here we have to put the object in the memo, not for efficiency or recursive reference issues, but for correctness:

```py
a = A()
f = a.f
g = a.g
c = C(f, g)
```

Dumping `c` will dump `a` twice, which means after load there will be two instances of `A()`. Now if `f` and `g` have side effects on `A`, in the above example, `c` is working on the same object, but after load it'd be working on two separate objects.

Reply:
This was I think the root of the recursive calls occuring with the `zipf` object mentioned earlier.

I can get a version of this being memoized tonight 👍

Reply:
Good catch Adrin. Let's also add a test for this.

Just to note that the existing memoization doesn't really support checking if an object was already encountered and retrieving this memoized object, as it wasn't necessary so far. So the memoization mechanism would have to be extended to support this. But I was expecting it would become necessary sooner or later.

Reply:
I don't have any good ideas. First thing that came to mind was storing a "breadcrumb" to the actual object in the memo dict and then if in `get_state_*` we find that we already saved that object's state, instead of saving it again, we just save the breadcrumb.

One disadvantage that comes to mind would be that during loading, we would have to load the objects in the exact same order, otherwise the breadcrumb would lead nowhere. There might be other issues, I haven't investigated deeply.

What was your idea?

Reply:
I've now implemented, so feel free to check out the new code in this branch, but essentially:

1) Store states of objects as normal in the `get_state`, but also store `__id__` of the original objects (this does mean objects that exist multiple times in the tree have their state stored multiple times, but I couldn't think of a tidy way to solve this without storing the state in multiple places)

2) Add a `LoadState` to get_instance calls to track `__id__` of objects as they get loaded

3) Check if an object with a given `__id__` has been instantiated already, if it has, return a pointer to that object instead of creating a new one.

I think storing in a JSON format makes it very hard to nicely store the state of an object that is referenced multiple times. Because like you said, we don't know which part of the tree will be loaded first, each time the object is referenced needs to be able to get the state. 

Perhaps a way to do this would be to do something along the lines of:

1) Store objects as normal, and memoize with `__id__`
2) If an object tries to get persisted a second times, remove state from original object, and leave `__id__` and `__is_memoized__` as a flag for each instance
3) At end of dump procedure, store each object that's been memoized like (2) in a `multi_instance_object` field outside normal tree, at top of JSON object
4) When loading, first check for any `multi_instance_object`s, and load them into the LoadState
5) Any object with the `_is_memoized_` flag just references the object in the LoadState

This would work I think, but it would be a pretty big shift from current save/load procedure...

Reply:
But on the plus side, the MVP simple version is passing a new test checking the pattern @adrinjalali mentioned above, with only one instance of `A()` 🥳

Reply:
I'll try to implement something that stores each object once done tomorrow, let me know if you think the procedure I described in steps 1-5 above sounds reasonable

Reply:
Wow, pretty big change. As always, big thanks @E-Aho for your excellent work.

I haven't done an in depth review, so just putting my thoughts out here.

1. I would have preferred this change to be its own PR.

I know that there is this potential bug that @adrinjalali mentioned, but maybe it's better to separate the memoization into its own PR and just add an `xfail`ing test for the bug? I think it's uncommon enough that we can survive with the main branch having this bug for a while.

2. A different signature for `get_instance_*` functions

Instead of adding a 3rd argument, it would be better if the `LoadState` holds the `src` as an additional attribute.

3. Deduplication is only done for method types

IIUC, the new memoization/dedup mechanism is only used for method types right now, is that right? I assume this is to keep the scope of the PR smaller, but if the whole mechanism is moved to a separate PR, I would like to see it applied for all types. That may mean that we should move it out of the individual `get_state_*` functions if possible.

4. Test too complex?

Maybe I'm missing something, but `test_works_when_given_multiple_bound_methods_attached_to_single_instance` seems to be doing more work than necessary to test for the bug. Instead of introducing `_MultipleMethodHolder`, would it not be sufficient to do something like this?

```python
foo = SomeObject()
estimator = FunctionTransformer(func=foo.bar, inverse_func=foo.baz)
```

Also, would this not be more correct?

```python
- assert loaded_obj.func1.__self__ == loaded_obj.func2.__self__
+ assert loaded_obj.func1.__self__ is loaded_obj.func2.__self__
```

5. Possible conflicts with Adrin's audit PR

@adrinjalali is working on auditing in #204. Since this PR introduces important changes to `get_instance`, there may be conflicts which may require coordination. Adrin can hopefully give more insights.

Reply:
I don't mind this work to move forward (memoization on load should be its own PR as @BenjaminBossan mentions though). I can handle the merge conflicts on my PR.

Reply:
> Perhaps a way to do this would be to do something along the lines of:
> ...
> 3. At end of dump procedure, store each object that's been memoized like (2) in a `multi_instance_object` field outside normal tree, at top of JSON object

Hmm. It feels wrong to me to have two classes of objects, those stored in the "tree" structure and those stored outside, in this new field. Yes, it would be nice to be able to avoid the dependency on the order in which we load objects, but maybe not at the cost of this added complexity.

Reply:
@BenjaminBossan thanks for putting up with this PR through it's so far many iterations! Has been a fun problem to think through.

> 1) I would have preferred this change to be its own PR.

Totally agree that loading memoization should be a seperate PR, it feels quite far out of scope for "Add method type in dispatch calls". I originally kept it in this branch because it was addressing an issue with the new functionality, but I'll revert to before the latest change, and open a new issue to track loading states for single objects.

> 2) A different signature for get_instance_* functions

Definitely much tidier to keep `get_instance`, also keeps closer to the pattern used in `SaveState`.

> 3) Deduplication is only done for method types
You are right as to why I only addressed `method` deduping, i.e to fix the problem Adrin mentioned, but if we were to implement this functionality in general, I agree it should cover any type.

> 4) Test too complex?
That would also work as a test, yes. I introduced the `_MultipleBoundMethodHolder` because to me it was initially clearer what was being tested, but you've changed my mind and I would be happy to use a test closer to what you suggested.